### PR TITLE
Implement InsertSocketPlugFree

### PR DESCRIPTION
**Pull request by bhollis**
This leaves the awa feature flag around and automatically limits actions depending on whether the action is free or you have the full AWA scope, just in case they ever want to give that out.

The practical effect of this is that when the manifest updates to make things free, this should automatically pick up the capability. Obviously we can't test it until Tuesday. Perks are already free so it goes ahead and tries it - I've put a cheeky little error message up that should go away after the deployment (I didn't bother making it translate-able but I will if folks think that's worthwhile).

### CODE REVIEW

File: src/app/bungie-api/rate-limit-config.ts [hunk 1]
```diff
@@ -30,4 +30,11 @@ export default function setupRateLimiter() {
   RateLimiterConfig.addLimiter(
     new RateLimiterQueue(/www\.bungie\.net\/Platform\/Destiny2\/Actions\/Items\/EquipItem/, 1, 100)
   );
+  RateLimiterConfig.addLimiter(
+    new RateLimiterQueue(
+      /www\.bungie\.net\/Platform\/Destiny2\/Actions\/Items\/InsertSocketPlugFree/,
+      2,
+      1100
```
Comment (bhollis) [inline src/app/bungie-api/rate-limit-config.ts:37]: From the release notes:

> The InsertSocketPlugFree API will initially be rate-limited to 2 socket actions per-second, per-user. This rate can be adjusted based on your feedback and service stability.

File: src/app/inventory/advanced-write-actions.ts [hunk 2]
```diff
@@ -25,39 +29,128 @@ let awaCache: {
...
+    // Must be reusable or randomized type
+    Boolean(
+      socket.socketDefinition.reusablePlugItems.length > 0 ||
+        socket.socketDefinition.reusablePlugSetHash ||
+        socket.socketDefinition.randomizedPlugSetHash
+    ) &&
+    // And have no cost to insert
+    !plug.plug?.insertionMaterialRequirementHash;
+
+  return free;
```
Comment (ryan-rushton) [inline src/app/inventory/advanced-write-actions.ts:68]: I think we need one more condition based on this comment

> - You will not be able to overwrite a 'non-free' plug item with a 'free' plug item (for example: replacing a Weapon mod with an 'Empty Socket Mod').

File: src/app/bungie-api/authenticated-fetch.ts [hunk 3]
```diff
@@ -74,18 +74,21 @@ async function responseIndicatesBadToken(response: Response) {
...
-      data.ErrorCode === PlatformErrorCodes.WebAuthRequired ||
-      // (also means the access token has expired)
-      data.ErrorCode === PlatformErrorCodes.WebAuthModuleAsyncFailed ||
-      data.ErrorCode === PlatformErrorCodes.AuthorizationRecordRevoked ||
-      data.ErrorCode === PlatformErrorCodes.AuthorizationRecordExpired ||
-      data.ErrorCode === PlatformErrorCodes.AuthorizationCodeStale ||
-      data.ErrorCode === PlatformErrorCodes.AuthorizationCodeInvalid)
-  );
+  try {
+    const data = await response.clone().json();
```
Comment (nev-r) [inline src/app/bungie-api/authenticated-fetch.ts:78]: LMAO i just finished adding the same try in a different copy of this
image

File: src/app/bungie-api/http-client.ts [hunk 4]
```diff
@@ -158,7 +158,10 @@ export function createHttpClient(
       credentials: withCredentials ? 'include' : 'omit',
     });
     const response = await fetchFunction(fetchOptions);
-    const data: ServerResponse<unknown> = await response.json();
+    let data: ServerResponse<unknown> | undefined;
+    try {
+      data = await response.json();
+    } catch {}
```
Comment (nev-r) [inline src/app/bungie-api/http-client.ts:164]: why is completely silent failure, preferable to throwing an error indicating that the response was invalid/not JSON?
i **really strongly disagree** with the major shift of allowing `HttpClient` to return undefined.
none of the bungie-api-ts functions account for this at all, they presume `HttpClient`'s `any` is their data type, assuming no error was thrown.

Comment (bhollis) [reply src/app/bungie-api/http-client.ts:164]: We can throw something if data doesn't end up returning anything - the core problem is the error thrown by parsing JSON causes this to skip checking for HTTP status codes.

**Review by nev-r — APPROVED** [review]

**Pull request merged**